[IMP] mail: Reply, Quotes, Expand & Forward ⤴⤵

mrsr-odoo opened the pull request:
Rationale

When having external recipients (customers, portal users), discussions from
and to Odoo is not always clear. Notably for internal users actual recipients
are unclear, as most are hidden in followers, and simple daily routines like
replying or forwarding content is not crystal clear.

We decided to move towards a flow more looking like email providers. Since
recipients are now better stored and proposed in Odoo (see [1] and [2]) we
add basics for Reply-All and Forward of messages inside Odoo. This allows to
ping / contact some people specifically, and better manipulate discussions.

Specification: Reply-All / Forward

Add two actions on comment messages in Odoo Chatter: Reply-All and Forward.

In Reply-All mode: fetch all non-followers recipients of replied message and
open a composer with all those people added in 'partner_ids'. This implies
creating partners for cc/to stored on message that are still not partners
(see [1]), and propose them based on chosen message (see [2]).

In Forward mode: transfer the content to contacts. This opens a composer with
original content. When posting, do not notify followers. Only specified
recipients have to be notified, thanks to recently introduced notification
option 'notify_skip_followers'.

Specification: notification control

Add a new 'notify_skip_followers' notification parameter, allowing to
skip followers in notification process.

Support it at composer level to have matching features in both standard
flow and composer-based flows.

Technical note: this is simply considered as posting a user_notification
for followers matter. Like this message type, followers are not fetched,
only asked recipients.

[1] see https://github.com/odoo/odoo/pull/191213 previously merged
[2] see https://github.com/odoo/odoo/pull/185240 previously merged

This PR belongs to a group of PRs tailored for having an email-like
discussion flow in documents
 * odoo/odoo#184824
 * odoo/odoo#187024
 * odoo/odoo#191213
 * odoo/odoo#185240
 * odoo/odoo#191358
 * odoo/odoo#186875
 * odoo/odoo#188642 : Final branch

Task-4273569: [mail] Reply, Quotes, Expand & Forward ⤴⤵
Prepares Task-4273479: [mail] Email-like recipients

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/186875.png)](https://mergebot.odoo.com/odoo/odoo/pull/186875)

thjo-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -496,6 +496,132 @@ export class Message extends Component {
+        const authorCheck = this.message.thread.suggestedRecipients.find((recipient) =>
+            recipient.email.includes(this.message.email_from)
+        );
+        const recipientEmails = [authorCheck.email];
+        const partners = await rpc("/mail/partner/from_email", {
+            emails: recipientEmails,
+            additional_values: { [authorCheck.email]: authorCheck.create_values || {} },
+        });
+        let author;
+        for (const index in partners) {
+            const partnerData = partners[index];
+            const persona = this.store.Persona.insert({ ...partnerData, type: "partner" });
+            const email = recipientEmails[index];
+            const recipient = this.message.thread.suggestedRecipients.find(
+                (recipient) => recipient.email === email
+            );
+            Object.assign(recipient, { persona });
+            author = persona;
+        }
```
Why do you need to do all of this? I have so much questions. If you have the email from your search above, couldn't you stop then? `recipientEmails` is a list containing a singular email and you never push anything inside it, why?

mrsr-odoo replied:
Hello
This function is especially for incoming emails, where message_type is `email`, there is no partner when there is an incoming email even in `suggestedRecipients` we only have name and email.
This function creates the partner or finds an existing partner (if any) and is used as `To:` in the reply-all dialog.
![image](https://github.com/user-attachments/assets/2d7ff544-e4b8-47ba-8e55-fbda3c6d918f)

mrsr-odoo replied:
>I get that this is what you are doing here, but why like this?
>
> recipientEmails is a list with only one email that's never updated, why a list then?

Actually `/mail/partner/from_email` accepts `emails` to be in the form of a list.
So I kept that part as a list. But I guess it is not readable
Alright making it simple now.

---
  > The day you have multiple partners for the same email you will encounter a crash in the loop. How do you handle this case?

Isn't that `_find_or_create_from_emails` returns a single partner for one email

thjo-odoo reviewed (changes requested):
@mrsr-odoo 
Hello, I already found some things that absolutely needs improvement.
And also had a lot of questions for your `fetchAuthorFromSuggestedRecipients`. 
This method doesn't make any sense past the half mark.

mrsr-odoo commented on addons/mail/static/src/core/common/message_actions.js:
```diff
@@ -88,6 +88,26 @@ messageActionsRegistry
+        onClick: (component) => {
+            component.onClickReplyMessage();
+        },
+        sequence: 78,
```
70 already taken for `set-new-message-separator`https://github.com/odoo/odoo/blob/2dde6694e849e2465e826fd0583668ee8f44b0ca/addons/mail/static/src/discuss/core/common/message_actions.js#L30
going with 71,72

mrsr-odoo commented on addons/mail/static/src/views/web/fields/html_composer_message_field/content_expandable_plugin.js:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    cleanForSave(root) {
+        root.querySelector("button")?.remove();
```
done
But somehow if I don't write this line it automatically removes the button from the body
Do you know how it works?

mrsr-odoo reviewed (commented):
Hello @thjo-odoo :slightly_smiling_face: 

Suggested changes are done 
and answered below

Thanks

thjo-odoo commented on addons/mail/static/src/views/web/fields/html_composer_message_field/content_expandable_plugin.js:
```diff
@@ -0,0 +1,49 @@
+            `<p><br/><button id="viewMore" contenteditable="false" class="badge rounded-pill border-0 m-1 py-0 px-1"><i class="fa fa-ellipsis-h fa-lg"/></button></p>`
+        );
+        const button = paragraph.querySelector("#viewMore");
+        this.addDomListener(button, "click", () => this.onClickViewButton());
```
You could also give it the bounded function instead of a brand new arrow function.

thjo-odoo reviewed (commented):
@mrsr-odoo 
Hello, here's some other comments for this one.

Cheers!

mrsr-odoo commented:
Hello @thjo-odoo :slightly_smiling_face: ,
Considering VIJI's comments on task pad 
The AI widget rewrites our reply content 
https://github.com/odoo/odoo/blob/41de88b930e569daea7624ee3655cef218f9136f/addons/mail/static/src/core/web/mail_composer_chatgpt.js#L25
So I updated the code as per need 
tried to add a button before the original message so it doesn't break the view

Also, I have updated the code for the scheduled messages as we have an edit option for that too.
  
----

> Also, from what I reread in the specs, shouldn't the reply all be sending a message to the same users as if I'd simply sent a message and opened the full composer?

first, we have to find a way to stop notifying all followers when in reply mode.
I tried to prevent followers for these messages in last commit (not sure if it is good or not ) :confused:

Please have a look 
Thanks

thjo-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -496,6 +496,138 @@ export class Message extends Component {
+            : "";
+        const datetime = this.message.date.toFormat("ccc, MMM d, yyyy 'at' hh:mm a");
+        const default_body = markup(
+            `<div class="o_mail_reply_content d-none">\
+            <span>On ${datetime} ${this.escape(author.name)} ${authorEmail} wrote </span>\
+            <blockquote>${getNonEditableMentions(this.message.body)}</blockquote></div>`
```
I have a thought for this, why not have the `d-none` be applied only on the span and blockquote, in a separate div, and have the button inside the `.o_mail_reply_content` before the span and blockquote.
```suggestion
            `<div class="o_mail_reply_content">
                <!-- Insert button here -->
                <div class="d-none">
                    <span>On ${datetime} ${this.escape(author.name)} ${authorEmail} wrote </span>
                    <blockquote>${getNonEditableMentions(this.message.body)}</blockquote>
                </div>
            </div>`
```

thjo-odoo commented on addons/mail/static/src/views/web/fields/html_composer_message_field/content_expandable_plugin.xml:
```diff
@@ -0,0 +1,11 @@
+    <t t-name="mail.ExpandableButton">
+        <br t-if="newLine"/>
+        <div contenteditable="false">
+            <button id="viewMore" class="badge rounded-pill border-0 m-1 py-0 px-1">
```
Could you instead of using id have a class so that you can identify the button more easily? Like `o-mail-Message-viewMore-btn` or something else if you have a better idea.

thjo-odoo reviewed (commented):
@mrsr-odoo
Hello, 
It seems GitHub decided to send my review before I was finished. So here's the second part.
I had a thought about a lot of issues you had and maybe we should add the button inside of the reply content. I added my explanations below if something is unclear feel free to send me a message.
Cheers!

mrsr-odoo commented on addons/mail/static/src/views/web/fields/html_composer_message_field/content_expandable_plugin.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates>
+    <t t-name="mail.ExpandableButton">
+        <br t-if="newLine"/>
```
we need an extra line before the content so we can write a message

mrsr-odoo commented:
Hello @thjo-odoo :wave: ,

Suggested changes are done 
and answered a few questions
Please have a look 
(As the validation already started could we move this task in Validation BE?)

Thanks :sparkles:

mrsr-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -496,6 +496,138 @@ export class Message extends Component {
+        const authorCheck = suggestedRecipients.find((recipient) =>
+            recipient.email.includes(this.message.email_from)
+        );
+        let author = authorCheck.persona;
```
Doesn't every message sent to odoo have author in suggestedRecipients or message.author?
If not then should we stop them further?

thjo-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -496,6 +496,142 @@ export class Message extends Component {
+        const authorCheck = suggestedRecipients.find((recipient) =>
+            recipient.email.includes(this.message.email_from)
+        );
+        let author = authorCheck?.persona;
```
Then this becomes a const.

thjo-odoo reviewed (commented):
@mrsr-odoo 
Here's another batch for this one.

mrsr-odoo commented:
Hello @thjo-odoo :wave: ,

Suggested changes are done.
Please have a look.

Thanks

mrsr-odoo commented on addons/mail/static/src/views/web/fields/html_composer_message_field/content_expandable_plugin.js:
```diff
@@ -0,0 +1,47 @@
+        const button = buttonTemplate.querySelector(".o-mail-Message-viewMore-btn");
+        this.addDomListener(button, "click", this.onClickViewButton);
+        ele.prepend(buttonTemplate);
+        this.config.onChange?.();
```
removed 
Previously used when  dom plugins was dependency

tde-banana-odoo commented:
@seb-odoo Would gladly have some input from you (and quite urgent, sorry for saying that, as we would like to have time to test the final prototype that depends on this branch).

seb-odoo reviewed (commented):
Hello, here is a quick code check as requested. I didn't test yet.

One important issue (markup check) and few code move/cleaning.

The PR mostly does data fetching and transforming in very specific places so it doesn't raise a lot of concern from us.

Disclaimer: I'm not familiar with full composer/editor code so I didn't really check that part.

seb-odoo commented on addons/mail/controllers/thread.py:
```diff
@@ -73,6 +73,23 @@ def mail_thread_messages(self, thread_model, thread_id, fetch_params=None):
             "messages": messages.ids,
         }
 
+    @http.route("/mail/thread/recipients", methods=["POST"], type="jsonrpc", auth="user")
+    def mail_thread_recipients(self, thread_model, thread_id, message_id=None, **kwargs):
+        thread = self._get_thread_with_access(thread_model, thread_id, mode='read', **kwargs)
+        if message_id:
+            message = self._get_message_with_access(message_id, mode="create", **kwargs)
```
Supporting `kwargs` here makes little sense if the function is limited to internal users (this is only used for portal stuff in other methods).

But how about project chatter from portal users? Is it taken into account at all in all the new tasks about composer? 
Portal should probably have specific route/overrides to turn possible recipients (limited to current followers of task/project)?

tde-banana-odoo replied:
Yes, had the same thoughts when adding the route. Still wondering if we want that feature for portal or not. Open question would say, makes sense in some cases (e.g. on a ticket, a portal user wants to add their boss, or pings their accounting person on a sale order), maybe not in other (i.e. don't spam partner table). Currently on hold.

seb-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -496,6 +520,121 @@ export class Message extends Component {
+    }
+}
+
+function getNonEditableMentions(body) {
```
Possibly move this function to utils/format to go with other similar body processing functions.

tde-banana-odoo commented:
@seb-odoo thanks for the check :) while looking more in-depth, I found indeed some code was unnecessary, and had to be improved with recent merges anyway. Will take into account while crash-reviewing :) .

tde-banana-odoo commented:
@robodoo override=ci/security

Reason: using innerHtml on previously markuped content, already validated as coming from message body. Then using marking on either validated content, either escaped. And anyway everything comes from mail.message body content, which is sanitized in DB.

tde-banana-odoo commented:
@robodoo r+ rebase-merge

Noted some post cleaning to perform (some code move), will be done in subsequent task that will improve the feature and make the final global features.

Bisous !

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.
